core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClientBuilder.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2018 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.circuitbreaker;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.function.Function;
+
+import javax.annotation.Nullable;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.client.Client;
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.common.Response;
+
+/**
+ * Builds a new {@link CircuitBreakerClient} or its decorator function.
+ *
+ * @param <T> the type of {@link CircuitBreakerClientBuilder}
+ * @param <U> the type of the {@link Client} that this builder builds or decorates
+ * @param <I> the type of outgoing {@link Request} of the {@link Client}
+ * @param <O> the type of incoming {@link Response} of the {@link Client}
+ */
+public abstract class CircuitBreakerClientBuilder<
+        T extends CircuitBreakerClientBuilder<T, U, I, O>, U extends CircuitBreakerClient<I, O>,
+        I extends Request, O extends Response> {
+
+    @Nullable
+    final CircuitBreakerStrategy strategy;
+
+    @Nullable
+    final CircuitBreakerStrategyWithContent<O> strategyWithContent;
+
+    CircuitBreakerMapping mapping = CircuitBreakerMapping.defaultMapping;
+
+    /**
+     * Creates a new builder with the specified {@link CircuitBreakerStrategy}.
+     */
+    CircuitBreakerClientBuilder(CircuitBreakerStrategy strategy) {
+        this(requireNonNull(strategy, "strategy"), null);
+    }
+
+    /**
+     * Creates a new builder with the specified {@link CircuitBreakerStrategyWithContent}.
+     */
+    CircuitBreakerClientBuilder(CircuitBreakerStrategyWithContent<O> strategyWithContent) {
+        this(null, requireNonNull(strategyWithContent, "strategyWithContent"));
+    }
+
+    private CircuitBreakerClientBuilder(@Nullable CircuitBreakerStrategy strategy,
```

Comment:
I just wanted to make a user specify either `RetryStrategy` or `RetryStrategyWithContent` when they create a builder so that only one of those strategies can be set.
I can do the same thing using setters and check the state inside the setters but I didn't want it because I thought it can confuse the user.
If you mind this constructor, I just can remove this private constructor and assign the variables in other constructors. What do you think?